### PR TITLE
[Checkpoint][2D][6/N] Add optimizer and update default_planner to core distributed

### PR DESCRIPTION
**Pull request by wz337**
This is the last PR for integrating 2D into core distributed. 

This PR does the following:
1. Add optimizer.py: this adds ability to load a state_dict in conjunction with FSDP sharded optimzer state. 
2. Update default_planner.py to support 2D checkpoint. 
3. Add test_fsdp_optim_state.py as a unit test for No. 1. 
4. Fix bug in torch/testing/_internal/distributed/checkpoint_utils.py
5. Rename the filename for the APIs that should be private. Will organize and cleanup further in following PRs. #90328

Docstring and integration test will be added in the following PRs.

### CODE REVIEW

File: test/distributed/checkpoint/test_fsdp_optim_state.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                storage_reader=dist_cp.FileSystemReader(CHECKPOINT_DIR),
+                # planner=ap.AdvLoadPlanner(),
+                planner=DefaultLoadPlanner(
+                    flatten_state_dict=True,
+                    flatten_sharded_tensors=True,
+                ),
+            )
+            model_2.load_state_dict(state_dict["model"])
+
+            optim_state = load_sharded_optimizer_state_dict(
```
Comment (wanchaol) [inline test/distributed/checkpoint/test_fsdp_optim_state.py:84]: shall we consolidate optimizer state dict to `load_state_dict` of dist_cp directly?

**Review by wanchaol — COMMENTED** [review]
Looking good, have some question about the planner init and I think we should probably have a plan about the user facing API consolidation.

File: torch/distributed/checkpoint/default_planner.py [hunk 2]
```diff
@@ -45,23 +64,63 @@
...
+        flatten_state_dict: bool = False,
+        flatten_sharded_tensors: bool = False,
+        dedup_replicated_tensors: bool = False,
+    ) -> None:
+        self.flatten_state_dict = flatten_state_dict
+        self.flatten_sharded_tensors = flatten_sharded_tensors
+        self.dedup_replicated_tensors = dedup_replicated_tensors
+        self.mappings = {}
+
+    def init(self, state_dict: STATE_DICT_TYPE, is_coordinator: bool) -> None:
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/checkpoint/default_planner.py:84]: Will fix this.

File: torch/distributed/checkpoint/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+                self.translation[wi.dest_index] = original_index
+
+            requests += reqs
+        return LoadPlan(requests)
+
+    def lookup_tensor(self, index: MetadataIndex) -> torch.Tensor:
+        return super().lookup_tensor(self.translation.get(index, index))
+
+
+def load_sharded_optimizer_state_dict(
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/checkpoint/optimizer.py:197]: Add issue to consolidate loading optimizer state dict to load_state_dict of dist_cp directly in #90325

File: test/distributed/checkpoint/test_fsdp_optim_state.py (hunk 1 above)
Comment (wz337) [reply test/distributed/checkpoint/test_fsdp_optim_state.py:84]: Filed a new issue to add test for this utils. #90327

File: torch/testing/_internal/distributed/checkpoint_utils.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def wrapper(self, *args: Tuple[object], **kwargs: Dict[str, Any]) -> None:
         # Only create temp_dir when rank is 0
         if dist.get_rank() == 0:
             temp_dir = tempfile.mkdtemp()
-            print(f"Using temp directory: {self.temp_dir }")
+            print(f"Using temp directory: {temp_dir}")
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/testing/_internal/distributed/checkpoint_utils.py:24]: Filed a new issue to add test for this utils. https://github.com/pytorch/pytorch/issues/90327

File: torch/distributed/checkpoint/default_planner.py [hunk 5]
```diff
@@ -242,3 +348,79 @@ def _create_default_local_metadata(state_dict: STATE_DICT_TYPE) -> Metadata:
     plan = _create_default_metadata_only_plan(state_dict)
     _, md = create_default_global_save_plan([plan])
     return md
+
+
+def _check_box_overlap(
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/checkpoint/default_planner.py:353]: File task to revisit what APIs should be public and what APIs should be private. https://github.com/pytorch/pytorch/issues/90328

File: torch/distributed/checkpoint/default_planner.py [hunk 6]
```diff
@@ -92,12 +151,42 @@ def transform_object(self, write_item: WriteItem, object: Any):
 
 
 class DefaultLoadPlanner(LoadPlanner):
+    """
+    LoadPlanner that adds multiple features on top of DefaultLoadPlanner.
```
Comment (wanchaol) [inline torch/distributed/checkpoint/default_planner.py:155]: is this comment accurate? shouldn't it be sth reverse?

```suggestion
    DefaultLoadPlanner that adds multiple features on top of LoadPlanner.
```

**Review by wanchaol — APPROVED** [review]
Looks good to me, thanks for addressing comments and tracking issues :)

File: torch/distributed/checkpoint/default_planner.py (hunk 6 above)
Comment (wz337) [reply torch/distributed/checkpoint/default_planner.py:155]: Thanks for catching.

**Pull request closed without merging**